bgcval2/functions/globalVolMean.py

```diff
@@ -0,0 +1,109 @@
+#
+# Copyright 2015, Plymouth Marine Laboratory
+#
+# This file is part of the bgc-val library.
+#
+# bgc-val is free software: you can redistribute it and/or modify it
+# under the terms of the Revised Berkeley Software Distribution (BSD) 3-clause license.
+
+# bgc-val is distributed in the hope that it will be useful, but
+# without any warranty; without even the implied warranty of merchantability
+# or fitness for a particular purpose. See the revised BSD license for more details.
+# You should have received a copy of the revised BSD license along with bgc-val.
+# If not, see <http://opensource.org/licenses/BSD-3-Clause>.
+#
+# Address:
+# Plymouth Marine Laboratory
+# Prospect Place, The Hoe
+# Plymouth, PL1 3DH, UK
+#
+# Email:
+# ledm@pml.ac.uk
+#
+"""
+.. module:: globalVolMean 
+   :platform: Unix
+   :synopsis: This function calculates the volume weighted mean of a field.
+
+.. moduleauthor:: Lee de Mora <ledm@pml.ac.uk>
+
+"""
+
+import numpy as np
+from ..bgcvaltools.dataset import dataset
+import os, sys
+import errno
+
+#from bgcvaltools.bgcvalpython import Area
+
+tmask     = 0
+pvol     = 0
+loadedArea = False
+
+def loadDataMask(gridfn):
+    """
+    Load data mask, heopfully only once!
+    """
+    global loadedArea
+    global tmask
+    global pvol
+    if not gridfn or not os.path.exists(gridfn):
+        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), gridfn)
+
+    nc = dataset(gridfn, 'r')        
+    try:
+        pvol   = nc.variables['pvol'][:]
+        tmask = nc.variables['tmask'][:]
+    except:
+        tmask = nc.variables['tmask'][:]
+        area = nc.variables['e2t'][:] * nc.variables['e1t'][:]
+        pvol = nc.variables['e3t'][:] * area
+        pvol = np.ma.masked_where(tmask==0, pvol)
+    nc.close()
+    loadedArea = True
```

Comment:
gah don't use globals, man, return values for local variables (more of a thing for the future, don't worry abt it here)

Reply:
So I try to not use globals , but I do think this is a justifiable use case for them. 

This function calculates the global volume mean, and it gets called on every netcdf file (one per year), so usually at least a hundred times for a model run. To do that calculation, we need to know the global total volume. Instead of reloading it for every single new netcdf file, this sets it as a global so it only needs to be loaded once!

Flipping FX files are always a nightmare though, right?!

Having said all that, I have a feeling that we're better off using the time variable version of volume, instead of one from the grid file. It might be possible to re-write this without globals.

Reply:
ah gotcha! That'll do in that case, performance always trumps style :car: